src/lsm/tree.zig

```diff
@@ -1079,35 +1065,31 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type) type {
             if (compacted_levels_odd) {
                 // Assert all visible tables haven't overflowed their max per level.
                 tree.manifest.assert_level_table_counts();
-
-                // Convert mutable table to immutable table for next bar.
-                tree.compact_mutable_table_into_immutable();
             }
         }
 
         /// Called after the last beat of a full compaction bar.
-        fn compact_mutable_table_into_immutable(tree: *Tree) void {
-            assert(tree.table_immutable.free);
-            assert((tree.compaction_op.? + 1) % constants.lsm_batch_multiple == 0);
-
-            if (tree.table_mutable.count() == 0) return;
+        fn swap_mutable_and_immutable(tree: *Tree, snapshot_min: u64) void {
+            assert(tree.table_immutable.mutability.immutable.flushed);
+            assert(snapshot_min > 0);
+            assert(snapshot_min < snapshot_latest);
 
-            // Sort the mutable table values directly into the immutable table's array.
-            const values_max = tree.table_immutable.values_max();
-            const values = tree.table_mutable.sort_into_values_and_clear(values_max);
-            assert(values.ptr == values_max.ptr);
+            assert((tree.compaction_op.? + 1) % constants.lsm_batch_multiple == 0);
 
             // The immutable table must be visible to the next bar.
             // In addition, the immutable table is conceptually an output table of this compaction
             // bar, and now its snapshot_min matches the snapshot_min of the Compactions' output
             // tables.
-            tree.table_immutable.reset_with_sorted_values(
-                snapshot_min_for_table_output(compaction_op_min(tree.compaction_op.?)),
-                values,
-            );
+            var current_table_mutable = tree.table_mutable;
+            current_table_mutable.make_immutable(snapshot_min);
+
+            var current_table_immutable = tree.table_immutable;
+            current_table_immutable.make_mutable();
+
+            tree.table_immutable = current_table_mutable;
+            tree.table_mutable = current_table_immutable;
 
```

Comment:
Since we asserted `tree.table_immutable.mutability.immutable.flushed` as a precondition, I think `tree.table_immutable.mutability == .mutable/.immutable` pre- and post- conditions are fair game.